File: 84-largest-rectangle-in-histogram/largest-rectangle-in-histogram.cpp
class Solution {
public:
    int largestRectangleArea(vector<int>& heights) {
        stack<int> st;
        int n = heights.size();
        int maxArea = 0;
        for(int i=0;i<=heights.size();i++)
        {
            while(!st.empty() && (i==n || heights[i]<heights[st.top()]))
            {
                int h = heights[st.top()];
                st.pop();
                int width = st.empty()?i:i-st.top()-1;
                maxArea = max(maxArea,h*width);
            }
            st.push(i);
        }

        return maxArea;
        
    }
};

File: 84-largest-rectangle-in-histogram/largest-rectangle-in-histogram.java
class Solution {
    public int largestRectangleArea(int[] heights) {
        int maxArea = 0;
        Stack<Integer> s = new Stack<>();
        for(int i=0;i<=heights.length;i++)
        {
            while(!s.isEmpty() && (i==heights.length || heights[s.peek()]>heights[i])){
                int h = heights[s.peek()];
                s.pop();
                int width = (s.isEmpty())? i :i-1- s.peek();
                maxArea = Math.max(maxArea,h*width);
            }
            s.push(i);
        }
         return maxArea;
    }
}